connector/failoverconnector/failover.go

```diff
@@ -0,0 +1,197 @@
+package failoverconnector
+
+import (
+	"errors"
+	"go.opentelemetry.io/collector/component"
+	"sync"
+	"time"
+)
+
+type consumerProvider[C any] func(...component.ID) (C, error)
+
+type failoverRouter[C any] struct {
+	consumerProvider consumerProvider[C]
+	cfg              *Config
+	pS               *pipelineSelector
+	consumers        []C
+	done             chan bool
+	inRetry          bool
+}
+
+var (
+	errNoValidPipeline = errors.New("All provided pipelines return errors")
+	errConsumer        = errors.New("Error registering consumer")
+)
+
+func newFailoverRouter[C any](provider consumerProvider[C], cfg *Config) *failoverRouter[C] {
+	return &failoverRouter[C]{
+		consumerProvider: provider,
+		cfg:              cfg,
+		pS:               newPipelineSelector(cfg),
+		done:             make(chan bool),
+	}
+}
+
+func (f *failoverRouter[C]) getCurrentConsumer() C {
+	return f.consumers[f.pS.currentIndex]
+}
+
+func (f *failoverRouter[C]) registerConsumers() error {
+	consumers := make([]C, 0)
+	for _, pipeline := range f.cfg.PipelinePriority {
+		newConsumer, err := f.consumerProvider(pipeline...)
+		if err == nil {
+			consumers = append(consumers, newConsumer)
+		} else {
+			return errConsumer
+		}
```

Comment:
nit: prefer to return early in order to allow for less nested logic
```suggestion
		if err != nil {
			return errConsumer
		}
		consumers = append(consumers, newConsumer)
```